x/subscription/types/cronspec.go

```diff
@@ -0,0 +1,396 @@
+		Minute: []CronItem{{Spec: &CronItem_Value{Value: 0}}},
+		Hour:   []CronItem{{Spec: &CronItem_Value{Value: 0}}},
+		Mday:   []CronItem{{Spec: &CronItem_Value{Value: 1}}},
+		Month:  []CronItem{{Spec: &CronItem_Value{Value: 1}}},
+		Wday:   []CronItem{{}},
+	}
+}
+
+func monthlyCronSpec() CronSpec {
+	return CronSpec{
+		Minute: []CronItem{{Spec: &CronItem_Value{Value: 0}}},
+		Hour:   []CronItem{{Spec: &CronItem_Value{Value: 0}}},
+		Mday:   []CronItem{{Spec: &CronItem_Value{Value: 1}}},
+		Month:  []CronItem{{}},
+		Wday:   []CronItem{{}},
+	}
+}
+
+func weeklyCronSpec() CronSpec {
+	return CronSpec{
+		Minute: []CronItem{{Spec: &CronItem_Value{Value: 0}}},
+		Hour:   []CronItem{{Spec: &CronItem_Value{Value: 0}}},
+		Mday:   []CronItem{{}},
+		Month:  []CronItem{{}},
+		Wday:   []CronItem{{Spec: &CronItem_Value{Value: 1}}},
+	}
+}
+
+func dailyCronSpec() CronSpec {
+	return CronSpec{
+		Minute: []CronItem{{Spec: &CronItem_Value{Value: 0}}},
+		Hour:   []CronItem{{Spec: &CronItem_Value{Value: 0}}},
+		Mday:   []CronItem{{}},
+		Month:  []CronItem{{}},
+		Wday:   []CronItem{{}},
+	}
+}
+
+func hourlyCronSpec() CronSpec {
+	return CronSpec{
+		Minute: []CronItem{{Spec: &CronItem_Value{Value: 0}}},
+		Hour:   []CronItem{{}},
+		Mday:   []CronItem{{}},
+		Month:  []CronItem{{}},
+		Wday:   []CronItem{{}},
+	}
+}
+
+// ParseCronSpec parse CronSpec from string
+func ParseCronSpec(s string) (CronSpec, error) {
+	switch s {
+	case "@yearly":
+		return yearlyCronSpec(), nil
+	case "@monthly":
+		return monthlyCronSpec(), nil
+	case "@weekly":
+		return weeklyCronSpec(), nil
+	case "@daily":
+		return dailyCronSpec(), nil
+	case "@hourly":
+		return hourlyCronSpec(), nil
+	}
+	parts := strings.Split(s, " ")
+	if len(parts) != 5 {
+		return CronSpec{}, errors.New("invalid cron spec")
+	}
+	minute, err := parseItemSpec(parts[0])
+	if err != nil {
+		return CronSpec{}, err
+	}
+	hour, err := parseItemSpec(parts[1])
+	if err != nil {
+		return CronSpec{}, err
+	}
+	mday, err := parseItemSpec(parts[2])
+	if err != nil {
+		return CronSpec{}, err
+	}
+	month, err := parseItemSpec(parts[3])
+	if err != nil {
+		return CronSpec{}, err
+	}
+	wday, err := parseItemSpec(parts[4])
+	if err != nil {
+		return CronSpec{}, err
+	}
+	return CronSpec{
+		Minute: minute,
+		Hour:   hour,
+		Mday:   mday,
+		Month:  month,
+		Wday:   wday,
+	}, nil
+}
+
+func parseItemSpec(s string) ([]CronItem, error) {
+	parts := strings.Split(s, ",")
+	var items = []CronItem{}
+	for _, part := range parts {
+		item, err := parseItemSpecSingle(part)
+		if err != nil {
+			return nil, err
+		}
+		items = append(items, item)
+	}
+	if len(items) == 0 {
+		return nil, errors.New("empty item spec")
+	}
+	return items, nil
+}
+
+func parseItemSpecSingle(s string) (CronItem, error) {
+	switch {
+	case s == "*":
+		return CronItem{}, nil
+	case strings.Contains(s, "-"):
+		parts := strings.Split(s, "-")
+		if len(parts) == 2 || len(parts) == 3 {
+			start, err := strconv.Atoi(parts[0])
```

Comment:
lgtm complains about these platform-specific functions -- could this use more precise / fixed with functions (`ParseUint`?)?

Reply:
done